### PR TITLE
Feat: Composite Design Pattern for installers

### LINKED ISSUE #212: Composite Design Pattern for composing Mono/Scriptable Objects Installers into tree structures
n/a

### LINKED ISSUE #215: UX: Organize Unity's Zenject menu to a more sensible order

### PR DESCRIPTION
**Pull request by tsgcpp**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] No compiler errors or warnings

## Pull request type

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## Issue Number

Issue Number: #211

*Create or search an issue here: [Extenject/Issues](https://github.com/svermeulen/Extenject/issues)*

## What is the current behavior?

- No changes against current behaviours

## What is the new behavior?

- Add "CompositeMonoInstaller" and "CompositeScriptableObjectInstaller" to "OptionalExtras". They are the composite pattern of Extenject installers
- Add a document about them named "CompositeInstaller.md", so I wonder if you could also read the md file
- Add tests for "CompositeMonoInstaller" and "CompositeScriptableObjectInstaller"
  - All asset files and prefabs are for the tests.
- Add editor scripts for "CompositeMonoInstaller" and "CompositeScriptableObjectInstaller" to warn circular references if any circular references are found in the installers

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

- You can see the components looks in "CompositeInstaller.md"

On which Unity version has this been tested?
--------------------------------------------
- [ ] 2020.4 LTS
- [ ] 2020.3
- [x] 2020.2
- [ ] 2020.1
- [x] 2019.4 LTS
- [ ] 2019.3
- [ ] 2019.2
- [ ] 2019.1
- [ ] 2018.4 LTS

**Scripting backend:**
- [x] Mono
- [x] IL2CPP

### CODE REVIEW

**Comment by Mathijs-Bakker** [comment]
@tsgcpp This is an awesome feature!
It's a lot to review, so it could take some days.
I like the tooling you've done and the circular dependency warning.

Great work!

File: UnityProject/Assets/Plugins/Zenject/OptionalExtras/CompositeInstaller/Tests/IntegrationTests/Install/TestCompositeMonoInstaller/FooInjectee.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+using UnityEngine;
+using Zenject;
+
+namespace Zenject.Tests.Installers.CompositeMonoInstallers
+{
+    public class FooInjectee
```
Comment (Mathijs-Bakker) [inline UnityProject/Assets/Plugins/Zenject/OptionalExtras/CompositeInstaller/Tests/IntegrationTests/Install/TestCompositeMonoInstaller/FooInjectee.cs]: Sorry, but could this be a typo? Did you mean FooInjector.cs?

Comment (Mathijs-Bakker) [reply UnityProject/Assets/Plugins/Zenject/OptionalExtras/CompositeInstaller/Tests/IntegrationTests/Install/TestCompositeMonoInstaller/FooInjectee.cs]: My bad... this is perfect grammar X)

**Review by Mathijs-Bakker — APPROVED** [review]
Well written clean code and tests!

**Comment by Mathijs-Bakker** [comment]
@tsgcpp This feature is a great addition to Zenject. 
I can merge your pr as submitted. Which means your feature will be in 'Optional Extras',
But it deserves to be standard functionality, so we could move it into into Zenject/source/...
What about it? Merge (for now) or move to the core?

**Comment by tsgcpp** [comment]
@Mathijs-Bakker I really appreciate your review and your suggestion.

> But it deserves to be standard functionality, so we could move it into into Zenject/source/...
What about it? Merge (for now) or move to the core?

If there is no problem, I would like to move them into "Zenject/source/...".
Would you tell me where to move them?

Is it ok to move them like this?
- https://github.com/tsgcpp/Extenject/compare/feature/composite-installer...tsgcpp:wip/feature/composite-installer_move_as_standard?expand=1
  - move scripts into "Zenject/source/..."
  - move tests and assets for them into "Zenject/OptionalExtras/IntegrationTests/..."
  - delete asmdef files for CompositeInstallers

**Comment by Mathijs-Bakker** [comment]
> Is it ok to move them like this?
>   
>   * move scripts into "Zenject/source/..."
>   * move tests and assets for them into "Zenject/OptionalExtras/IntegrationTests/..."
>   * delete asmdef files for CompositeInstallers

Yes exactly.
You have all paths correct.

**Comment by tsgcpp** [comment]
@Mathijs-Bakker OK, I've moved composite installers and their tests.
Could you please confirm my new commits?

**Review by Mathijs-Bakker — APPROVED** [review]

**Review by Mathijs-Bakker — APPROVED** [review]

File: Documentation/CompositeInstaller.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+- [CompositeInstaller](#compositeinstaller)
+- [Table Of Contents](#table-of-contents)
+- [Introduction](#introduction)
+  - [CompositeMonoInstaller](#compositemonoinstaller)
+  - [CompositeScriptableObjectInstaller](#compositescriptableobjectinstaller)
+  - [FYI](#fyi)
+
+</details>
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
```
Comment (Mathijs-Bakker) [inline Documentation/CompositeInstaller.md:16]: This piece of code is generated by a GitHub action. (DocToc)
These lines (1 to 14) can be moved to to under the ``## Table Of Contents`` header without causing any issues.

File: Documentation/CompositeInstaller.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+## CompositeInstaller
+
+## Table Of Contents
+- [CompositeMonoInstaller](#compositemonoinstaller)
+- [CompositeScriptableObjectInstaller](#compositescriptableobjectinstaller)
+
```
Comment (Mathijs-Bakker) [inline Documentation/CompositeInstaller.md]: Line 19 and 20 can be removed.
When the DocToc generated lines are moved to here.

File: UnityProject/Assets/Plugins/Zenject/Source/Install/CompositeScriptableObjectInstaller.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
+using System.Collections.Generic;
+using UnityEngine;
+using Zenject;
+
+namespace Zenject
+{
+    [CreateAssetMenu(fileName = "CompositeScriptableObjectInstaller", menuName = "Installers/CompositeScriptableObjectInstaller")]
```
Comment (Mathijs-Bakker) [inline UnityProject/Assets/Plugins/Zenject/Source/Install/CompositeScriptableObjectInstaller.cs]: This ``CreateAssetMenu`` line can be removed 
after a new ``[MenuItem("Assets/Create/Zenject/...]`` is created (in Editor/ZenMenuItems.cs)

**Review by Mathijs-Bakker — CHANGES_REQUESTED** [review]
Cool, 
Now we decided to move this feat to Zenject's core we also want to rearrange the ``CompositeScriptableObjectInstaller`` menu item from Installers to Zenject. 

Schermafbeelding 2021-02-07 om 12 43 12

This menu is generated in ``Zenject/Source/Editor/ZenMenuItems.cs``.
Where we can add something like:
```cs
[MenuItem("Assets/Create/Zenject/Composite Scriptable Object Installer", false, 1)]
public static void CreateCompositeScriptableObjectInstaller()
{
    AddCSharpClassTemplate("Composite Scriptable Object Installer", "UntitledInstaller",
        "using UnityEngine;"
        + "\nusing Zenject;"
        // -->
}
```
For a better UX it make sense to reorder the Installer section.
So it reads like this:
- Installer
- Mono Installer
- Scriptable Object Installer
- Composite Scriptable Object Installer
        
May I ask you to take care of this?
Drop me a line for any help.

Cheers!

**Comment by tsgcpp** [comment]
@Mathijs-Bakker Now, I'm editing "ZenMenuItems".

screenshot 2021-02-07 234750

I have a question.
Should I add a separator between other installers and "Composite Scriptable Object Installer" like this?

image

`Installer`, `Mono Installer`, `Scriptable Object Installer` will create a "script" file, but "Composite Scriptable Object Installer" will create an "asset" file of "CompositeScriptableObjectInstaller".

Would you give me your opinion?

## WIP diff
https://github.com/tsgcpp/Extenject/compare/feature/composite-installer...tsgcpp:wip/feature/composite-installer?expand=1

**Comment by Mathijs-Bakker** [comment]
> @Mathijs-Bakker Now, I'm editing "ZenMenuItems".
> 
> screenshot 2021-02-07 234750
> 
> I have a question.
> Should I add a separator between other installers and "Composite Scriptable Object Installer" like this?
> 
> image
> 

Without the line separator looks the cleanest.

That was also my initial concern to reverse the order of the installers. 
- Installer
- Mono Installer
- Scriptable Object Installer
- Composite Scriptable Object Installer

It is clearer for the user to grasp the differences between the installers. Don't you think?

**Comment by tsgcpp** [comment]
@Mathijs-Bakker Thank you for your opinion.

> It is clearer for the user to grasp the differences between the installers. Don't you think?

Yes, I think so.
OK, I will edit like this:
screenshot 2021-02-07 234750

Could you wait for a few days?
I will complete it and fix the document.

**Comment by Mathijs-Bakker** [comment]
> Could you wait for a few days?
> I will complete it and fix the document.

No worries!
Take your time!

**Comment by tsgcpp** [comment]
@Mathijs-Bakker I've added  a menu item "Composite Scriptable Object Installer" and fixed "Table Of Contents" of "CompositeInstaller.md".

I also edited "UnityInspectorListEditor" to warn circular references in any contexts like this (not pushed yet):
image

Should I push this ([diff](https://github.com/tsgcpp/Extenject/compare/14e5750...fbc3f3c?expand=1))?

**Comment by Mathijs-Bakker** [comment]
Yeah, you can push this diff to this pr.
Then we have everything in place. @tsgcpp 
👍

**Comment by tsgcpp** [comment]
@Mathijs-Bakker I've pushed the diff.
~~Could you please confirm new commits?~~

**Comment by tsgcpp** [comment]
@Mathijs-Bakker 
I'm sorry. Can you give me more time?
It seems that there is a bug when Extenject validation are executed.
Then, Unity freezes.

**Comment by tsgcpp** [comment]
I found the cause of the bug.
If there is a circular references in any composite installers, "StackOverflowException" is not sometimes thrown and Unity is terminated.
I should have added a validation to composite installers before their "InstallBindings"...

I will push new commits to fix that when it is resolved.

**Comment by tsgcpp** [comment]
OK, I've resolved the bug and added assertions to composite installers before "InstallBindings" of their leaves.
If any circular references are found in composite installers, "ZenjectException" will be thrown in Runtime and "Zenject -> Validate ...".
I also fixed validation methods to reduce allocations because the validation methods became used in Runtime (No allocations up to 4 chains).

Could you please confirm new commits again? @Mathijs-Bakker

**Review by Mathijs-Bakker — APPROVED** [review]
@tsgcpp
Good! I see you have added the asserts.
Great!

Runned this locally (2019.LTS)... everything behaved as expected. 
Awesome work!

I will merge this pr tomorrow. 

The readme.md needs a reference to the new doc too. And a brief explanation about this feature.
I can append a section to it.

:star: :star: :star: Awesome work! :star: :star: :star:

**Comment by Mathijs-Bakker** [comment]
Merged.

**Comment by tsgcpp** [comment]
I really appreciate your supports! @Mathijs-Bakker 
I hope this contribution will help the developers.

**Pull request merged**